src/compiler/checker.ts

```diff
@@ -17011,11 +17014,38 @@ namespace ts {
                 }
             }
 
-            function isIdenticalTo(source: Type, target: Type): Ternary {
+            function areUnionsOrIntersectionsTooLarge(source: Type, target: Type, reportErrors: boolean) {
+                if ((source.flags & TypeFlags.UnionOrIntersection) && (target.flags & TypeFlags.UnionOrIntersection)) {
+                    const sourceUnionOrIntersection = source as UnionOrIntersectionType;
+                    const targetUnionOrIntersection = target as UnionOrIntersectionType;
+
+                    if (sourceUnionOrIntersection.objectFlags & targetUnionOrIntersection.objectFlags & ObjectFlags.PrimitiveUnion) {
+                        // There's a fast path for comparing primitive unions
+                        return false;
+                    }
+
+                    const sourceSize = sourceUnionOrIntersection.types.length;
+                    const targetSize = targetUnionOrIntersection.types.length;
+                    if (sourceSize * targetSize > 1E7) {

```

Comment:
I think the limit should be 1000000 (1M). That's the limit we use for non-primitive unions in subtype reduction.

Reply:
Will do.